File: services/typescript/koa/src/routes/handlers/patch.ts
import type { RouterContext } from "@koa/router";
import { type State, parseId, parsePatchedTask } from "app";
import { getLogger } from "telemetry";
import { AppErrors, sendError } from "../errors.js";

export function patchHandler(state: State) {
  return async (ctx: RouterContext) => {
    const logger = getLogger();

    const id = parseId(ctx.params.id);
    if (!id.success) {
      logger.warn({ method: ctx.req.method, path: ctx.path }, "Invalid path");
      return sendError(ctx, AppErrors.InvalidPath);
    }

    const patchedTask = parsePatchedTask(ctx.request.body);
    if (!patchedTask.success) {
      logger.warn(
        { method: ctx.req.method, path: ctx.path },
        "Invalid body JSON"
      );
      return sendError(ctx, AppErrors.InvalidJsonBody);
    }

    const task = state.tasks.get(id.data);
    if (!task) {
      logger.warn(
        { id: id.data, method: "PATCH", path: ctx.path },
        "Task not found"
      );
      return sendError(ctx, AppErrors.TaskNotFound);
    }

    const previousOperation = task.operation;
    task.operation = patchedTask.data.operation;

    logger.info(
      {
        id: id.data,
        secret: task.secret.length,
        from_operation: previousOperation.toString().toLowerCase(),
        to_operation: task.operation.toString().toLowerCase(),
        method: "PATCH"
      },
      "Patched task"
    );

    ctx.status = 200;
    ctx.body = task;
  };
}

File: services/python/starlette/src/routes/handlers/patch.py
import structlog
import json
from uuid import UUID
from pydantic import ValidationError
from starlette.requests import Request
from starlette.responses import Response, JSONResponse

from app.params import parse_patched_task
from app.state import AppState
from app.task import PatchedTask, Task
from app.operation import Operation
from routes.errors import send_error, AppErrors


async def patch_handler(request: Request) -> Response:
    logger = structlog.get_logger()

    id: UUID = request.path_params["id"]

    try:
        body = await request.json()
        patched_task: PatchedTask = parse_patched_task(body)
    except (json.JSONDecodeError, ValidationError):
        logger.warn("Invalid body JSON", method=request.method, path=request.url.path)
        return send_error(AppErrors.InvalidJsonBody)

    state: AppState = request.app.state.app_state

    async with request.app.state.tasks_lock:
        task: Task | None = state.tasks.get(id)

        if task is None:
            logger.warn(
                "Task not found",
                id=str(id),
                method=request.method,
                path=request.url.path,
            )
            return send_error(AppErrors.TaskNotFound)

        previous_operation: Operation = task.operation

        state.tasks[id].operation = patched_task.operation

    logger.info(
        "Patched task",
        id=str(id),
        secret=len(task.secret),
        from_operation=previous_operation.lower(),
        to_operation=patched_task.operation.lower(),
        method=request.method,
    )

    return JSONResponse(task.model_dump())
